File: 589-n-ary-tree-preorder-traversal/589-n-ary-tree-preorder-traversal.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    List<Integer> ans = new ArrayList<Integer>();
    public void preorderUtil(Node root) {
        if (root != null) {
            ans.add(root.val);
            for (Node child: root.children) {
                preorderUtil(child);
            }
        }
    }
    public List<Integer> preorder(Node root) {
        preorderUtil(root);
        return ans;
    }
}

File: 589-n-ary-tree-preorder-traversal/589-n-ary-tree-preorder-traversal.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {Node|null} root
 * @return {number[]}
 */
let ans = []
var preorderUtil = function(root) {
    if (root) {
        ans.push(root.val)
        root.children.forEach((ele) => preorderUtil(ele))
    }
}
var preorder = function(root) {
    ans.length = 0
    preorderUtil(root)
    return ans
};